### PR TITLE
upgrade to do all tables, not just system tables

### LINKED ISSUE #562: Add option to auto upgrade user tables
Hi,
from todays point, it is possible to automatically run `mariadb-upgrade --upgrade-system-tables` on container start, by setting `MARIADB_AUTO_UPGRADE` to `TRUE`.

Unfortunately, this is not covering my user tables, which leads to the fact, that theres still a human interaction required, to complete the MariaDB upgrade.

I am sure, that this is not the expected behavior for important production deployments, but for smaller projects/home use, it would be very nice, to have an optional way/variable, to also upgrade the user tables on container startup.

From my recent lookup in the current `docker-entrypoint.sh`, this can may be handled with only one additional if condition - but i am not sure, whether there is any greater impact or something different, that must be kept in mind, while implementing this feature.

If the maintainers are fine with a feature like that, i may also contribute it via PR.

### PR DESCRIPTION
**Pull request by grooverdan**
Might get some startup timeout on those that have a huge number of tables during upgrade.

WDYT @WhoAmI0501

Closes #562

### CODE REVIEW

**Comment by grooverdan** [comment]
appears to be 11.0+ errors - https://github.com/grooverdan/mariadb-docker/actions/runs/8014222826

**Comment by WhoAmI0501** [comment]
Hi and sorry for the late response.

What kind of timeoutes are we talking about? Docker Health Checks?

**Comment by grooverdan** [comment]
Delay starting - https://jira.mariadb.org/browse/MDEV-27060

While bitnami --force was the cause of this the measures of upgrade time are still relevant.

**Comment by WhoAmI0501** [comment]
> Delay starting - https://jira.mariadb.org/browse/MDEV-27060
> 
> While bitnami --force was the cause of this the measures of upgrade time are still relevant.

I read the bug story multiple times and still dont really get it. The official MariaDB container should fix a problem with the entrypoint script of a inofficial third party container?

Well... on the one hand, imho this shouldn't be our problem here. On the other hand, its already fixed and it may be kind of asocial towards to the bitnami community, to remove the fix now.

What about adding a additional environment variable, to control, whether user tables should be also covered and leaving the system table parameter as the default behavior?

**Comment by WhoAmI0501** [comment]
Just seen that the bug story is also a few years old.

So i checked the current Bitnami repository and didnt found any `--force` parameter in the scripts (but may i also overseen something). You may also have a look on here: https://github.com/bitnami/containers/tree/main/bitnami/mariadb/11.2/debian-12

**Comment by grooverdan** [comment]
Ignore the `--force` aspect. Because of the large number of user tables, 2hrs where required to perform the upgrade in this user's case.

> What about adding a additional environment variable, to control, whether user tables should be also covered and leaving the system table parameter as the default behavior?

While it solves the immediate there are two cases (lots of user tables, and those that don't know), the remaining user question is how does a user ever now they need to upgrade their user tables?

There's an aspect that its not a new problem and providing an env variable is just like putting a `upgrade-system-tables` (or its inverse) in a configuration file, and abandonment of responsibility to the user. Its like the opt in version problem described in the issue, you don't want to find X years later that the version is incompatible (unlikely but mistakes happen when not tested).

So in short the competing goals are:
* achieve a quick startup time
* make users not worry their user tables are getting neglected in mariadb-upgrade

I'd rather:
a) the server covered the mariadb-upgrade of non-system tables automaticly because inline takes far too long.
b) optimizing / parallelizing of mariadb-upgrade of non-system tables to get down to a reasonable time to work (probably more achievable)

**Comment by WhoAmI0501** [comment]
I absolutely understand and support your point of view, as well as your goals.

>  b) optimizing / parallelizing of mariadb-upgrade of non-system tables to get down to a reasonable time to work (probably more achievable)

As we already read, theres sometimes an disk I/O bottleneck, which leads the upgrade task to take very long. So based on that fact, i guess, that parallelizing may wouldnt make it better in any way.

In either way, it seems like this topic should be discussed in the MariaDB repository/project itself? Unfortunately, i cannot contribute anything than my view as a user to this problem, since i never had any contact with the code of MariaDB and its tooling.

**Comment by grooverdan** [comment]
> I absolutely understand and support your point of view, as well as your goals.
> 
> > b) optimizing / parallelizing of mariadb-upgrade of non-system tables to get down to a reasonable time to work (probably more achievable)
> 
> As we already read, theres sometimes an disk I/O bottleneck, which leads the upgrade task to take very long. So based on that fact, i guess, that parallelizing may wouldnt make it better in any way.

It reads a lot of tables definitions but doesn't dive into the entire table contents.

>> In either way, it seems like this topic should be discussed in the MariaDB repository/project itself?

yep.

Thankfully @vaintroub provided me with a [reference implementation](https://github.com/mariaDB/server/commit/4532dae0164ba6bfbc0086cdfa0110b890989e05).

> Unfortunately, i cannot contribute anything than my view as a user to this problem, since i never had any contact with the code of MariaDB and its tooling.

Your view as a user is valuable enough :-)
